### PR TITLE
feat: Allow route render methods to return a function

### LINKED ISSUE #125: Nested `<Route />` being rendered despite parent `<Route />` rendering `null`
Hey all. Love the library. Great work.

Having an issue with a complex usage of the Redux-connected `createConnectedRouter` (although I sense my issue is agnostic of the fact that I'm using that as opposed to the "basic" usage).

**My implementation:**
```javascript
function App({ children }) {
  return (
    <div>
      <h1>The App</h1>
      {children}
    </div>
  )
}

function NestedComponent() {
  return <div>The nested component</div>
}

function getData() {
  return promiseReturningFunction().then((someValue) => { someValue })
}

const routeConfig = makeRouteConfig(
  <Route
    path='/'
    Component={App}
    getData={getData}
    render={({ Component, data, props }) => {
      if (!data) {
        return null
      }

      return <Component {...props} />
    }}
  >
    <Route path='/nested' Component={NestedComponent} />
  </Route>
)

const createStoreWithEnhancers = composeEnhancers(
  createHistoryEnhancer({
    protocol: new BrowserProtocol(),
    middlewares: [queryMiddleware],
  }),
  createMatchEnhancer(new Matcher(routeConfig)),
)(createStore)

const reducer = combineReducers({
  found: foundReducer,
})

const store = createStoreWithEnhancers(reducer)
const ConnectedRouter = createConnectedRouter({
  render: createRender({}),
})

ReactDOM.render(
  <Provider store={store}>
    <ConnectedRouter resolver={resolver} />
  </Provider>
)
```

With the code above, I am noticing that on the initial render pass, although the `render` for the top-level `<Route />` is returning `null` (as a result of `getData`), the child `<Route />`'s component (`NestedComponent`) is still getting rendered for a brief moment (without the wrapping `App` component being rendered). Note that the render lifecycle of `App` isn't even being invoked. The `NestedComponent` is rendering completely outside of that context. In the second pass that occurs after `getData` is resolved, both components render as expected (with `NestedComponent` being rendered as a child of `App`).

When tracing the unexpected render of `NestedComponent` it looks like it's simply coming from the `resolveMatch` of the `BasicRouter`.

My expectation would be that child `<Route />`s and their `Component`s do not render if their parent `<Route />`'s `render` is returning `null`. Am I perhaps thinking about this the wrong way?

Happy to provide more debugging information if necessary.

**Edit**
For some added clarification, the visual output is as follows.
1. The page simply reads "The nested component" for the first pass (unexpected).
2. After `getData` is resolved, the page reads (as expected) "The App" followed by "The nested component"

### PR DESCRIPTION
**Pull request by taion**
Closes #125

### CODE REVIEW

**Comment by taion** [comment]
not sure if we should deprecate the old thing

File: src/ElementsRenderer.js [hunk 1]
```diff
@@ -26,10 +29,14 @@ function accumulateElement(children, element) {
...
     });
 
-    return React.cloneElement(element, groups);
+    return element instanceof Function
+      ? element(groups)
+      : React.cloneElement(element, groups);
   }
 
-  return React.cloneElement(element, { children });
+  return element instanceof Function
```
Comment (jquense) [inline src/ElementsRenderer.js]:
```suggestion
  return typeof element === 'function'
```

File: src/ElementsRenderer.js [hunk 2]
```diff
@@ -26,10 +29,14 @@ function accumulateElement(children, element) {
       groups[groupName] = groupElements.reduceRight(accumulateElement, null);
     });
 
-    return React.cloneElement(element, groups);
+    return element instanceof Function
```
Comment (jquense) [inline src/ElementsRenderer.js]:
```suggestion
  return typeof element === 'function'
```

File: src/ElementsRenderer.js [hunk 3]
```diff
@@ -3,16 +3,19 @@ import React from 'react';
...
+      PropTypes.element,
+      PropTypes.func,
+    ]),
   ).isRequired,
 };
 
 function accumulateElement(children, element) {
   if (!children) {
-    return element;
+    return element instanceof Function ? element() : element;
```
Comment (taion) [inline src/ElementsRenderer.js]:
```suggestion
    return typeof element === 'function' ? element() : element;
```

**Comment by taion** [comment]
not going to deprecate the old API for now (among other things, the asymmetry between group and single child behavior that i think makes the API slightly nicer when writing these by hand also makes thing a bit more annoying to deal with things in an automated manner

**Pull request merged**